### PR TITLE
Math: Add Auditory functions library

### PR DESCRIPTION
**Pull request by singalsu**
The PR contains one commit for the library and another commit for cmocka tests.

### CODE REVIEW

**Comment by singalsu** [comment]
It's still work in progress, the Mel filterbank apply needs a test.

**Comment by lgirdwood** [comment]
@ShriramShastry pls can you review. Thanks !

**Comment by ShriramShastry** [comment]
> @ShriramShastry pls can you review. Thanks !

Sure, I 'II review the PR.
Thank you!!

**Comment by singalsu** [comment]
I just updated the PR with unit tests for Mel filterbank with various configurations:

```
sof/build_ut$ ctest --verbose -R auditory
UpdateCTestConfiguration  from :/home/singalsu/work/current/sof/sof/build_ut/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/singalsu/work/current/sof/sof/build_ut/DartConfiguration.tcl
Test project /home/singalsu/work/current/sof/sof/build_ut
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 52
    Start 52: auditory

52: Test command: /home/singalsu/work/current/sof/sof/build_ut/test/cmocka/src/math/auditory/auditory
52: Test timeout computed to be: 10000000
52: 1..10
52: Max absolute error =  3.00 (max  5.00), error RMS =  0.96 (max  3.00)
52: ok 1 - test_mel_filterbank_16_test1
52: Max absolute error =  3.00 (max  5.00), error RMS =  0.88 (max  3.00)
52: ok 2 - test_mel_filterbank_32_test1
52: Max absolute error =  1.00 (max  5.00), error RMS =  0.47 (max  3.00)
52: ok 3 - test_mel_filterbank_16_test2
52: Max absolute error =  1.00 (max  5.00), error RMS =  0.51 (max  3.00)
52: ok 4 - test_mel_filterbank_32_test2
52: Max absolute error =  1.00 (max  5.00), error RMS =  0.32 (max  3.00)
52: ok 5 - test_mel_filterbank_16_test3
52: Max absolute error =  1.00 (max  5.00), error RMS =  0.32 (max  3.00)
52: ok 6 - test_mel_filterbank_32_test3
52: Max absolute error =  3.00 (max  5.00), error RMS =  1.22 (max  3.00)
52: ok 7 - test_mel_filterbank_16_test4
52: Max absolute error =  3.00 (max  5.00), error RMS =  1.37 (max  3.00)
52: ok 8 - test_mel_filterbank_32_test4
52: Max absolute error =  1.00 (max  1.50), error RMS =  0.22 (max  0.50)
52: ok 9 - test_hz_to_mel
52: Max absolute error =  3.00 (max  5.00), error RMS =  1.30 (max  1.50)
52: ok 10 - test_mel_to_hz
52: # ok - tests
1/1 Test #52: auditory .........................   Passed    0.07 sec

The following tests passed:
	auditory

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   0.07 sec

```

**Comment by singalsu** [comment]
The previous push with FFT patches included passed CI, so I removed them from the latest push. CI will pass after #6061 is merged.

File: src/include/sof/math/auditory.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+ *
+ * Triangles data Number triangles is mel_bins
+ * then for each triangle:
+ *	index to next triangle 0, start fft_bin 1, length of this triangle 2,
+ *	triangle weight values 3..N
+ */
+struct psy_mel_filterbank {
+	int32_t log_mult; /**< Out, QX.Y scale for log, log10, or dB format */
+	int32_t scale; /**< Out, Scale Q16.16 used for Q1.15 triangles */
+	int32_t scale_inv; /**< Out, Inverse scale Q16.16 applied to triangles data */
```
Comment (singalsu) [inline src/include/sof/math/auditory.h:39]: Fields scale and scale_inv can be converted to local variables. The correction for Mel triangles gain is done with scale_log2 now.

**Comment by singalsu** [comment]
New push fixed merge conflict in test/cmocka/src/math/CMakeLists.txt, no other changes.

File: test/cmocka/src/math/auditory/auditory.c [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+#define HZ_TO_MEL_MAX_ERROR_RMS  0.5
+#define MEL_TO_HZ_MAX_ERROR_ABS  5.0
+#define MEL_TO_HZ_MAX_ERROR_RMS  1.5
+
+#define MEL_FB16_MAX_ERROR_ABS  5.0
+#define MEL_FB16_MAX_ERROR_RMS  3.0
+#define MEL_FB32_MAX_ERROR_ABS  5.0
+#define MEL_FB32_MAX_ERROR_RMS  3.0
+
+#define DEBUGFILES
```
Comment (singalsu) [inline test/cmocka/src/math/auditory/auditory.c:39]: This needs to be #undef

File: test/cmocka/src/math/auditory/ref_auditory.m [hunk 3]
```diff
@@ -0,0 +1,194 @@
+% ref_auditory - Generate C header files for auditory library unit tests
+
+% SPDX-License-Identifier: BSD-3-Clause
+%
+% Copyright(c) 2022 Intel Corporation. All rights reserved.
+
+function ref_auditory()
+
+% get_ref_hz_to_mel('ref_hz', 'ref_mel', 'ref_revhz', 'ref_hz_to_mel.h');
```
Comment (singalsu) [inline test/cmocka/src/math/auditory/ref_auditory.m:9]: This line should not be commented to generate the test data

File: src/math/auditory/auditory.c [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+/* For scaling log2 power to log, log10, or dB*/
+#define ONE_OVER_LOG2E_Q29	Q_CONVERT_FLOAT(0.6931471806, 29) /* 1/log2(exp(1))) */
+#define ONE_OVER_LOG2TEN_Q29	Q_CONVERT_FLOAT(0.3010299957, 29) /* 1/log2(10) */
+#define TEN_OVER_LOG2TEN_Q29	Q_CONVERT_FLOAT(3.0102999566, 29) /* 10/log2(10) */
+
+ /* TODO: Used for now to compensate FFT level vs. reference Matlab code for 512 long
+  * FFT. Check FFT lib scaling and make a closed form formula for all FFT lengths.
+  */
+#define FFT_LEVEL_OFFSET_Q16 Q_CONVERT_FLOAT(18.0, 16)
```
Comment (singalsu) [inline src/math/auditory/auditory.c:42]: Remove, this is no more used.

File: src/include/sof/math/auditory.h [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+#ifndef __SOF_MATH_AUDITORY_H__
+#define __SOF_MATH_AUDITORY_H__
+
+#include <sof/math/fft.h>
+#include <sof/math/log.h>
+#include <stdint.h>
+
+#define AUDITORY_EPS_Q31	1 /* Smallest nonzero Q1.31 value */
+#define AUDITORY_LOG_2P31_Q26	Q_CONVERT_FLOAT(21.4875625974, 26) /* log(2^31) */
```
Comment (singalsu) [inline src/include/sof/math/auditory.h:18]: No more used

File: src/math/auditory/auditory.c [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+{
+	uint32_t log_arg;
+	int32_t log;
+	int16_t mel;
+
+	if (hz < 0)
+		return 0;
+
+	/* Log function input is unsigned UQ32.0, output is UQ5.27 */
+	//log_arg = (((int32_t)hz << 16) / 700 + (1L << 16)) << 10; /* Scale by 2^26 */
```
Comment (singalsu) [inline src/math/auditory/auditory.c:62]: Some old code version in comment, remove

File: src/math/auditory/auditory.c [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+#include <sof/math/auditory.h>
+#include <sof/math/decibels.h>
+#include <sof/math/fft.h>
+#include <sof/math/log.h>
+#include <sof/math/numbers.h>
+#include <ipc/topology.h>
+#include <errno.h>
+#include <stdint.h>
+
+#define EPS_Q31 1
```
Comment (singalsu) [inline src/math/auditory/auditory.c:18]: Non used #define

File: src/math/auditory/auditory.c [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+#define TWO_Q29 Q_CONVERT_FLOAT(2, 29)
+
+#define LOG_SCALE_HZMEL_Q26 Q_CONVERT_FLOAT(18.02182669, 26)
+#define LOG_MULT_HZMEL_Q20 Q_CONVERT_FLOAT(1126.9941805389, 20)
+#define ONE_OVER_HZDIV_Q31 Q_CONVERT_FLOAT(1.0 / 700, 31)
+#define ONE_OVER_MELDIV_Q31 Q_CONVERT_FLOAT(1.0 / 1126.9941805389, 31)
+
+#define MEL_MAX_Q2 Q_CONVERT_FLOAT(4358.351140, 2) /* 32767 Hz */
+
+#define LOG_2P31_Q26 Q_CONVERT_FLOAT(21.4876, 26)
```
Comment (singalsu) [inline src/math/auditory/auditory.c:31]: This is not used

**Review by singalsu — COMMENTED** [review]
Found some non-used definitions from old code version

**Comment by lgirdwood** [comment]
@wszypelt @lrudyX this PR shouldn't fail on the DMIC test as it's not executed yet on that path. Can you check and confirm if OK to merge. Thanks !

File: src/math/auditory/auditory.c [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+/**
+ * Convert Hz to Mel
+ * Wikipedia https://en.wikipedia.org/wiki/Mel_scale
+ * mel = 1126.9941805389 * log(1 + hz / 700)
+ *
+ * @param hz Frequency as Q16.0 Hz, max 32767 Hz
+ * @return Value as Q14.2 Mel, max 4358.4 Mel
+ */
+int16_t psy_hz_to_mel(int16_t hz)
```
Comment (lgirdwood) [inline src/math/auditory/auditory.c:45]: qualify for static inline i.e. will it get called a lot in hot code ?

**Review by lgirdwood — COMMENTED** [review]
LGTM, but lets confirm if some funcs need to be static inline.

**Comment by wszypelt** [comment]
@lgirdwood I ran the CI tests again

**Comment by lgirdwood** [comment]
@singalsu CI looks good, but some conflicts.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
license server issues..... re run the CI

File: test/cmocka/src/math/auditory/ref_mel_filterbank_16_test1.h [hunk 10]
```diff
@@ -0,0 +1,557 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2022 Intel Corporation. All rights reserved.
+ */
+
+/* This file was created with script ref_auditory.m */
```
Comment (singalsu) [inline test/cmocka/src/math/auditory/ref_mel_filterbank_16_test1.h:6]: I'll do the same update as I did for matrix cmocka test vectors export. Would path "sof/test/cmocka/m" be a suitable for common Matlab language helper functions?

Comment (singalsu) [reply test/cmocka/src/math/auditory/ref_mel_filterbank_16_test1.h]: Done now

File: src/math/auditory/auditory.c (hunk 9 above)
Comment (singalsu) [reply src/math/auditory/auditory.c:45]: No, it's used only initialization. The Mel filterbank apply is the hot part. Did you find other inline candidates?

Comment (lgirdwood) [reply src/math/auditory/auditory.c:45]: no, but these only came to attention as they were small and simple.

**Comment by singalsu** [comment]
Reviewers, can you please review this.

File: test/cmocka/src/math/auditory/mfcc_get_mel_filterbank.m [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+		state.log_scale = 1;
+	case 'MEL_LOG10'
+		state.log_scale = 1/log(10);
+	case 'MEL_DB'
+		state.log_scale = 10/log(10);
+	otherwise
+		error('Invalid mel_log');
+end
+
+if 0
```
Comment (ShriramShastry) [inline test/cmocka/src/math/auditory/mfcc_get_mel_filterbank.m:72]: can turn off them using `if isequal (variable_enable, flag)`

File: test/cmocka/src/math/auditory/mfcc_mel_to_hz.m [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+% SPDX-License-Identifier: BSD-3-Clause
+%
+% Copyright (c) 2022, Intel Corporation. All rights reserved.
+
+function hz = mfcc_mel_to_hz(mel)
+
+hz = 700 * (exp(mel / 1126.9941805389) - 1);
+
+end
```
Comment (ShriramShastry) [inline test/cmocka/src/math/auditory/mfcc_mel_to_hz.m:13]: new line error ?

File: src/include/sof/math/auditory.h [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+	int16_t *scratch_data2; /**< Scratch, Packed triangles data */
+	int16_t *data; /**< Out, Packed triangles data */
+	int scratch_length1; /**< In, Length of first scratch */
+	int scratch_length2; /**< In, Length of second scratch */
+	int fft_bins; /**< In, Number of FFT bins */
+	int half_fft_bins; /**< In, fft_bins / 2 + 1 */
+	int mel_bins; /**< In, Number of Mel frequency bins */
+	int data_length; /**< Out, Number of int16_t words in triangles data */
+	enum psy_mel_log_scale mel_log_scale; /**< In, LOG, LOG10 or DB to select Mel format */
+	bool slaney; /**< In, Apply Slaney type normalization for Mel filterbank if true */
```
Comment (ShriramShastry) [inline src/include/sof/math/auditory.h:51]: can rename `slaney`  flag?

**Review by ShriramShastry — COMMENTED** [review]
Code looks sound to me !!

**Comment by andrula-song** [comment]
LGTM~

File: src/include/sof/math/auditory.h (hunk 13 above)
Comment (singalsu) [reply src/include/sof/math/auditory.h:51]: How, would slaney_normalize be better?

File: test/cmocka/src/math/auditory/mfcc_mel_to_hz.m (hunk 12 above)
Comment (singalsu) [reply test/cmocka/src/math/auditory/mfcc_mel_to_hz.m:13]: yep it was missing

File: test/cmocka/src/math/auditory/mfcc_get_mel_filterbank.m (hunk 11 above)
Comment (singalsu) [reply test/cmocka/src/math/auditory/mfcc_get_mel_filterbank.m]: Done, in beginning now plot_triangles and plot_matlab_triangles to enable the plots.

**Comment by singalsu** [comment]
> LGTM~

Don't hesitate to use the vote green "V" and red "X" buttons since you took the effort of reading the code!

**Comment by singalsu** [comment]
SOFCI TEST

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**